flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mailbox.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks.mailbox;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A mailbox is basically a blocking queue for inter-thread message exchange in form of {@link Runnable} objects between
+ * multiple producer threads and a single consumer.
+ */
+public interface Mailbox extends MailboxReader, MailboxWriter {
```

Comment:
I propose to abstract the message type.
Instead of using `Runnable`, make it a generic parameter (`Mailbox<Message>, MailboxReader<Message>, MailboxWriter<Message>`).

Reply:
Why?

Reply:
It eases reasoning about `Mailbox` abstraction. So far we don't have any contracts/expectation for message type.
I hope, that using generic type here would not impact performance.

Reply:
Generalizing it should not bloat `Mailbox` api and implementation at all, imo.

Here you are using `Runnable` as "message", it's rather implementation detail choice.
Technically, `StreamTask` could have been using plain data objects as messages (using the same `Mailbox` abstraction). Potentially, such approach simplifies reasoning about control flow.